src/tokenserver/auth/browserid.rs

```diff
@@ -0,0 +1,486 @@
+use actix_web::{web, Error};
+use async_trait::async_trait;
+use reqwest::{blocking::Client as ReqwestClient, StatusCode};
+use serde::{de::Deserializer, Deserialize, Serialize};
+
+use super::VerifyToken;
+use crate::tokenserver::{
+    error::{ErrorLocation, TokenserverError},
+    settings::Settings,
+};
+
+use core::time::Duration;
+use std::{convert::TryFrom, sync::Arc};
+
+/// The information extracted from a valid BrowserID assertion.
+#[derive(Clone, Debug, Default, Deserialize, PartialEq)]
+pub struct VerifyOutput {
+    pub device_id: Option<String>,
+    pub email: String,
+    pub generation: Option<i64>,
+    pub keys_changed_at: Option<i64>,
+}
+
+/// The verifier used to verify BrowserID assertions.
+#[derive(Clone)]
+pub struct RemoteVerifier {
+    audience: String,
+    issuer: String,
+    fxa_verifier_url: String,
+    // We need to use an Arc here because reqwest's blocking client doesn't use one internally,
+    // and we want to share keep-alive connections across threads
+    request_client: Arc<ReqwestClient>,
```

Comment:
The async request client _does_ use an Arc internally, but I specifically chose to use the blocking one. We don't want these BrowserID requests to FxA happening on Actix's worker threads; instead, I use `web::block` to invoke the requests on threads in the blocking threadpool. If I had used the async client, I would have had to wrap each request invocation with `web::block(futures::executor::block_on(|| { .. }))` instead of simply using `web::block`.

Reply:
> We don't want these BrowserID requests to FxA happening on Actix's worker threads;

Why not? The `verify` method is async so it seems we could use the async `reqwest` client -- `.await`ing it instead of the `block`/`block_on` combo.

Reply:
Ah I think I had myself confused. Since the requests are async, it's okay if they run on the worker threads since they wouldn't be blocking them. Thanks for the catch